java-checks/src/main/java/org/sonar/java/checks/NullPointerCheck.java

```diff
@@ -453,138 +433,97 @@ void registerAssignedSymbol(Symbol symbol) {
-    }
-
-    void mergeConditionalAnd(ConditionalState leftConditionalState, ConditionalState rightConditionalState) {
-      // copies the learned values to the parent.
-      trueState.copyValuesFrom(leftConditionalState.trueState);
-      trueState.copyValuesFrom(rightConditionalState.trueState);
-      // invalidates both false states and copies values to the parent
-      falseState.copyValuesFrom(leftConditionalState.falseState.invalidateValues());
-      falseState.copyValuesFrom(rightConditionalState.falseState.invalidateValues());
-    }
-
-    void mergeConditionalOr(ConditionalState leftConditionalState, ConditionalState rightConditionalState) {
-      // invalidates both true states and copies value to the parent.
-      trueState.copyValuesFrom(leftConditionalState.trueState.invalidateValues());
-      trueState.copyValuesFrom(rightConditionalState.trueState.invalidateValues());
-      // copies the learned values to the parent.
-      falseState.copyValuesFrom(leftConditionalState.falseState);
-      falseState.copyValuesFrom(rightConditionalState.falseState);
-    }
-  }
-
-  @VisibleForTesting
-  static class State {
-    @Nullable
-    final State parentState;
-    final Map<VariableSymbol, AbstractValue> variables;
-
-    public State() {
-      this.parentState = null;
-      this.variables = new HashMap<>();
-    }
-
-    public State(State parentState) {
-      this.parentState = parentState;
-      this.variables = new HashMap<>();
+  public class NullPointerVisitor extends SymbolicExecutionCheck {
+    @Override
+    protected void onArrayAccess(ExecutionState executionState, ArrayAccessExpressionTree tree) {
+      checkForIssue(tree.expression(), MESSAGE_NULLABLE_EXPRESSION, MESSAGE_NULL_LITERAL);
     }
 
-    // returns the value of the variable in the current state.
-    public AbstractValue getVariableValue(VariableSymbol variable) {
-      for (State state = this; state != null; state = state.parentState) {
-        AbstractValue result = state.variables.get(variable);
-        if (result != null) {
-          return result;
+    @Override
+    protected void onExecutableElementInvocation(ExecutionState executionState, Tree tree, List<ExpressionTree> arguments) {
+      if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+        MethodInvocationTree methodInvocation = (MethodInvocationTree) tree;
+        Symbol symbol = methodInvocation.symbol();
+        if (symbol.isMethodSymbol()) {
+          MethodJavaSymbol methodSymbol = (MethodJavaSymbol) symbol;
+          List<JavaSymbol> parameters = methodSymbol.getParameters().scopeSymbols();
+          if (!parameters.isEmpty()) {
+            for (int i = 0; i < methodInvocation.arguments().size(); i += 1) {
+              // in case of varargs, there could be more arguments than parameters. in that case, pick the last parameter.
+              if (checkNullity(parameters.get(i < parameters.size() ? i : parameters.size() - 1)) == NullableState.NOTNULL) {
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)